### ISSUE TITLE
Copying from an empty default list pointer to a list of data attempts to `copy_nonoverlapping` from a null pointer

### ISSUE THREAD

**Issue by ObsidianMinor** [capnproto/capnproto-rust#525]
In `wire_helpers::set_list_pointer`, copying between data lists triggers an unconditional `copy_nonoverlapping` even if there's nothing to copy from. This triggers a precondition check in std which causes a panic in debug mode. [Note that even if the effectively copied size (count * size_of::<T>()) is 0, the pointers must be non-null and properly aligned.](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html#safety)

**Comment by dwrensha** [comment]
Thanks for the report! I'll work on a fix for this.

Here's a test case that fails:
```rust
    #[test]
    fn set_data_list() {
        use crate::test_capnp::test_all_types;

        let mut message = message::Builder::new_default();
        let root: test_all_types::Builder<'_> = message.init_root();

        let mut message2 = message::Builder::new_default();
        let mut root2 : test_all_types::Builder<'_> = message2.init_root();
        root2.set_data_list(root.into_reader().get_data_list().unwrap()).unwrap();
    }
```

```
thread 'tests::set_data_list' panicked at core/src/panicking.rs:221:5:
unsafe precondition(s) violated: ptr::copy_nonoverlapping requires that both pointer arguments are aligned and non-null and the specified memory ranges do not overlap
```

**Comment by dwrensha** [comment]
Fixed in 04d2dd6e4a17a44f2b10119e58697a09a067f439.
It would be good to audit our other usages of  `ptr::copy_nonoverlapping()` too.

**Comment by marmeladema** [comment]
Would it be possible to backport it to 0.19.x series?

**Comment by dwrensha** [comment]
I fixed a few more instances of the bug in https://github.com/capnproto/capnproto-rust/commit/7b094f0deae3bb1c0efe7c51b8f094113992c133.

I've pushed capnp-v0.20.2 and capnp-v0.19.8 to crates.io with the fix.

**Comment by marmeladema** [comment]
Thank you very much!
